README.md
# unclogged

[![Build Status](https://travis-ci.org/RackSec/unclogged.svg?branch=master)](https://travis-ci.org/RackSec/unclogged)

[![Clojars Project](https://img.shields.io/clojars/v/unclogged.svg)](https://clojars.org/unclogged)

Easy, performant syslog for Clojure, exposed as a [manifold][manifold]
stream. Supports [RFC 3164][RFC3164] (classic BSD syslog), and the
[RFC5424][RFC5424] compliant [RFC 5426][RFC5426] (syslog over UDP) and
[RFC 6587][RFC6587] (syslog over TCP and TLS) specifications.

Since all of this is exposed using the stream abstractions from
[manifold][manifold], it works well with a variety of synchronous and
asynchronous programs, including [`core.async`][coreasync].

[manifold]: https://www.github.com/ztellman/manifold
[RFC3164]: http://tools.ietf.org/html/rfc3164
[RFC5424]: http://tools.ietf.org/html/rfc5424
[RFC5426]: http://tools.ietf.org/html/rfc5426
[RFC6587]: http://tools.ietf.org/html/rfc5426
[coreasync]: https://github.com/clojure/core.async

## Motivation

A couple of reasons you might want this:

- By having the API be a stream, your code becomes much easier to test. No
  need to set up a real syslog server; just see what comes out the other end.
- You can also use Clojure-like shorthand for severities, facilities and the
  like, for example, you can use `:info` (short for `:informational`, the name
  used in the spec, which also works) or even `"INFO"` instead of having to
  remember that that's severity 6.
- You can log strings or structured messages; they'll be serialized
  appropriately regardless.
- Breadth of different specs and transports (UDP, TCP, TLS) supported.

## Usage

TODO

## License

Copyright © 2016 Rackspace Hosting, Inc.

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# 0.9.0
- Adds ability to customize buffer size when creating a sink (#7)

# 0.9.1
- Fix release so it is deployed from origin, not from fork


test/unclogged/core_test.clj
(ns unclogged.core-test
  (:require
   [clojure.test :as t :refer [deftest testing is are]]
   [unclogged.core :as c]
   [manifold.stream :as s]
   [taoensso.timbre :refer [info spy]])
  (:import
   [com.cloudbees.syslog.sender TcpSyslogMessageSender UdpSyslogMessageSender]
   [com.cloudbees.syslog Facility Severity MessageFormat SyslogMessage]
   [java.io CharArrayWriter]))

(defn facility-bit-flag
  "In syslog, facilities don't use the lowest 2 bits; they're ints
  shifted left by 3."
  [n]
  (bit-shift-left n 3))

(defn compose-message
  [int]
  {:message (str "only in message: " int)
   :message-id (str "message id: " int)})

(deftest facility-tests
  (testing "original enum is a fixed point"
    (are [facility]  (= facility (#'unclogged.core/parse-facility facility))
      Facility/KERN
      Facility/USER
      Facility/MAIL
      Facility/DAEMON
      Facility/AUTH
      Facility/SYSLOG
      Facility/LPR
      Facility/NEWS
      Facility/UUCP
      Facility/CRON
      Facility/AUTHPRIV
      Facility/FTP
      Facility/NTP
      Facility/AUDIT
      Facility/ALERT
      Facility/CLOCK
      Facility/LOCAL0
      Facility/LOCAL1
      Facility/LOCAL2
      Facility/LOCAL3
      Facility/LOCAL4
      Facility/LOCAL5
      Facility/LOCAL6
      Facility/LOCAL7))
  (testing "from numerical codes"
    (are [code facility] (let [flag (facility-bit-flag code)]
                           (= facility (#'unclogged.core/parse-facility flag)))
      0 Facility/KERN
      1 Facility/USER
      2 Facility/MAIL
      3 Facility/DAEMON
      4 Facility/AUTH
      5 Facility/SYSLOG
      6 Facility/LPR
      7 Facility/NEWS
      8 Facility/UUCP
      9 Facility/CRON
      10 Facility/AUTHPRIV
      11 Facility/FTP
      12 Facility/NTP
      13 Facility/AUDIT
      14 Facility/ALERT
      15 Facility/CLOCK
      16 Facility/LOCAL0
      17 Facility/LOCAL1
      18 Facility/LOCAL2
      19 Facility/LOCAL3
      20 Facility/LOCAL4
      21 Facility/LOCAL5
      22 Facility/LOCAL6
      23 Facility/LOCAL7))
  (testing "from facility labels"
    (are [s facility] (= (#'unclogged.core/parse-facility s) facility)
      "KERN" Facility/KERN
      "USER" Facility/USER
      "MAIL" Facility/MAIL
      "DAEMON" Facility/DAEMON
      "AUTH" Facility/AUTH
      "SYSLOG" Facility/SYSLOG
      "LPR" Facility/LPR
      "NEWS" Facility/NEWS
      "UUCP" Facility/UUCP
      "CRON" Facility/CRON
      "AUTHPRIV" Facility/AUTHPRIV
      "FTP" Facility/FTP
      "NTP" Facility/NTP
      "AUDIT" Facility/AUDIT
      "ALERT" Facility/ALERT
      "CLOCK" Facility/CLOCK
      "LOCAL0" Facility/LOCAL0
      "LOCAL1" Facility/LOCAL1
      "LOCAL2" Facility/LOCAL2
      "LOCAL3" Facility/LOCAL3
      "LOCAL4" Facility/LOCAL4
      "LOCAL5" Facility/LOCAL5
      "LOCAL6" Facility/LOCAL6
      "LOCAL7" Facility/LOCAL7))
  (testing "from lower case"
    (are [s facility] (= facility (#'unclogged.core/parse-facility s))
      "kern" Facility/KERN
      "user" Facility/USER
      "mail" Facility/MAIL
      "daemon" Facility/DAEMON
      "auth" Facility/AUTH
      "syslog" Facility/SYSLOG
      "lpr" Facility/LPR
      "news" Facility/NEWS
      "uucp" Facility/UUCP
      "cron" Facility/CRON
      "authpriv" Facility/AUTHPRIV
      "ftp" Facility/FTP
      "ntp" Facility/NTP
      "audit" Facility/AUDIT
      "alert" Facility/ALERT
      "clock" Facility/CLOCK
      "locaL0" Facility/LOCAL0
      "locaL1" Facility/LOCAL1
      "locaL2" Facility/LOCAL2
      "locaL3" Facility/LOCAL3
      "locaL4" Facility/LOCAL4
      "locaL5" Facility/LOCAL5
      "locaL6" Facility/LOCAL6
      "locaL7" Facility/LOCAL7))
  (testing "from keyword"
    (are [s facility] (= facility (#'unclogged.core/parse-facility s))
      :kern Facility/KERN
      :user Facility/USER
      :mail Facility/MAIL
      :daemon Facility/DAEMON
      :auth Facility/AUTH
      :syslog Facility/SYSLOG
      :lpr Facility/LPR
      :news Facility/NEWS
      :uucp Facility/UUCP
      :cron Facility/CRON
      :authpriv Facility/AUTHPRIV
      :ftp Facility/FTP
      :ntp Facility/NTP
      :audit Facility/AUDIT
      :alert Facility/ALERT
      :clock Facility/CLOCK
      :locaL0 Facility/LOCAL0
      :locaL1 Facility/LOCAL1
      :locaL2 Facility/LOCAL2
      :locaL3 Facility/LOCAL3
      :locaL4 Facility/LOCAL4
      :locaL5 Facility/LOCAL5
      :locaL6 Facility/LOCAL6
      :locaL7 Facility/LOCAL7)))

(deftest severity-tests
  (testing "original enum is a fixed point"
    (are [severity] (= severity (#'unclogged.core/parse-severity severity))
      Severity/ALERT
      Severity/CRITICAL
      Severity/ERROR
      Severity/WARNING
      Severity/NOTICE
      Severity/INFORMATIONAL
      Severity/DEBUG))
  (testing "from numerical codes"
    (are [code severity] (= severity
                            (#'unclogged.core/parse-severity code))
      1 Severity/ALERT
      2 Severity/CRITICAL
      3 Severity/ERROR
      4 Severity/WARNING
      5 Severity/NOTICE
      6 Severity/INFORMATIONAL
      7 Severity/DEBUG))
  (testing "from labels"
    (are [label severity] (= severity
                             (#'unclogged.core/parse-severity label))
      "ALERT" Severity/ALERT
      "CRITICAL" Severity/CRITICAL
      "ERROR" Severity/ERROR
      "WARNING" Severity/WARNING
      "NOTICE" Severity/NOTICE
      "INFORMATIONAL" Severity/INFORMATIONAL
      "DEBUG" Severity/DEBUG))
  (testing "from lower case labels"
    (are [label severity] (= severity
                             (#'unclogged.core/parse-severity label))
      "alert" Severity/ALERT
      "critical" Severity/CRITICAL
      "error" Severity/ERROR
      "warning" Severity/WARNING
      "notice" Severity/NOTICE
      "informational" Severity/INFORMATIONAL
      "debug" Severity/DEBUG))
  (testing "from keywords"
    (are [kw severity] (= severity
                          (#'unclogged.core/parse-severity kw))
      :alert Severity/ALERT
      :critical Severity/CRITICAL
      :error Severity/ERROR
      :warning Severity/WARNING
      :notice Severity/NOTICE
      :informational Severity/INFORMATIONAL
      :debug Severity/DEBUG))
  (testing "from aliases"
    (are [sev-alias severity] (let [parse #'unclogged.core/parse-severity
                                    as-str (name sev-alias)
                                    upper-case (.toUpperCase ^String as-str)]
                                (= severity
                                   (parse sev-alias)
                                   (parse as-str)
                                   (parse upper-case)))
      :info Severity/INFORMATIONAL
      :err Severity/ERROR
      :warn Severity/WARNING)))

(deftest message-format-tests
  (testing "from strings"
    (are [s fmt] (= fmt (#'unclogged.core/parse-message-format s))
      "RFC 3164" MessageFormat/RFC_3164
      "RFC-3164" MessageFormat/RFC_3164
      "RFC_3164" MessageFormat/RFC_3164
      "RFC3164" MessageFormat/RFC_3164
      "rfc 3164" MessageFormat/RFC_3164
      "rfc-3164" MessageFormat/RFC_3164
      "rfc_3164" MessageFormat/RFC_3164
      "rfc3164" MessageFormat/RFC_3164

      "RFC 5424" MessageFormat/RFC_5424
      "RFC-5424" MessageFormat/RFC_5424
      "RFC_5424" MessageFormat/RFC_5424
      "RFC5424" MessageFormat/RFC_5424
      "rfc 5424" MessageFormat/RFC_5424
      "rfc-5424" MessageFormat/RFC_5424
      "rfc_5424" MessageFormat/RFC_5424
      "rfc5424" MessageFormat/RFC_5424))
  (testing "from keywords"
    (are [kw fmt] (= fmt (#'unclogged.core/parse-message-format kw))
      :RFC-3164 MessageFormat/RFC_3164
      :RFC_3164 MessageFormat/RFC_3164
      :RFC-3164 MessageFormat/RFC_3164
      :RFC_3164 MessageFormat/RFC_3164
      :RFC3164 MessageFormat/RFC_3164
      :rfc-3164 MessageFormat/RFC_3164
      :rfc_3164 MessageFormat/RFC_3164
      :rfc3164 MessageFormat/RFC_3164

      :RFC-5424 MessageFormat/RFC_5424
      :RFC_5424 MessageFormat/RFC_5424
      :RFC-5424 MessageFormat/RFC_5424
      :RFC_5424 MessageFormat/RFC_5424
      :RFC5424 MessageFormat/RFC_5424
      :rfc-5424 MessageFormat/RFC_5424
      :rfc_5424 MessageFormat/RFC_5424
      :rfc5424 MessageFormat/RFC_5424)))

(deftest ->syslog-msg-tests
  (testing "no defaults, all keys"
    (let [severity (#'unclogged.core/parse-severity :info)
          facility (#'unclogged.core/parse-facility :kern)
          contents {:message "hello"
                    :message-id "xyzzy"
                    :app-name "unclogged"
                    :hostname "ditka"
                    :process-id "1234"
                    :severity severity
                    :facility facility}
          syslog-msg (#'unclogged.core/->syslog-msg {} contents)]
      (is (= "hello" (.toString (.getMsg ^SyslogMessage syslog-msg))))
      (is (= "xyzzy" (.getMsgId ^SyslogMessage syslog-msg)))
      (is (= "unclogged" (.getAppName ^SyslogMessage syslog-msg)))
      (is (= "ditka" (.getHostname ^SyslogMessage syslog-msg)))
      (is (= "1234" (.getProcId ^SyslogMessage syslog-msg)))
      (is (= severity (.getSeverity ^SyslogMessage syslog-msg)))
      (is (= facility (.getFacility ^SyslogMessage syslog-msg)))))
  (testing "some defaults, all keys"
    (let [severity (#'unclogged.core/parse-severity :info)
          facility (#'unclogged.core/parse-facility :kern)
          defaults {:app-name "unclogged"
                    :hostname "ditka"
                    :process-id "1234"}
          contents {:message "hello"
                    :message-id "xyzzy"
                    :severity severity
                    :facility facility}
          syslog-msg (#'unclogged.core/->syslog-msg defaults contents)]
      (is (= "hello" (.toString (.getMsg ^SyslogMessage syslog-msg))))
      (is (= "xyzzy" (.getMsgId ^SyslogMessage syslog-msg)))
      (is (= "unclogged" (.getAppName ^SyslogMessage syslog-msg)))
      (is (= "ditka" (.getHostname ^SyslogMessage syslog-msg)))
      (is (= "1234" (.getProcId ^SyslogMessage syslog-msg)))
      (is (= severity (.getSeverity ^SyslogMessage syslog-msg)))
      (is (= facility (.getFacility ^SyslogMessage syslog-msg)))))
  (testing "type coercions"
    (let [defaults {:app-name "unclogged"
                    :hostname "ditka"
                    :process-id 1234}
          contents {:message [[:a] [[[{:b :c}]]]]
                    :message-id "xyzzy"
                    :severity :info
                    :facility :kern}
          syslog-msg (#'unclogged.core/->syslog-msg defaults contents)]
      (is (= "[[:a] [[[{:b :c}]]]]"
             (.toString ^CharArrayWriter (.getMsg ^SyslogMessage syslog-msg))))
      (is (= "xyzzy" (.getMsgId ^SyslogMessage syslog-msg)))
      (is (= "unclogged" (.getAppName ^SyslogMessage syslog-msg)))
      (is (= "ditka" (.getHostname ^SyslogMessage syslog-msg)))
      (is (= "1234" (.getProcId ^SyslogMessage syslog-msg)))
      (is (= Severity/INFORMATIONAL
             (.getSeverity ^SyslogMessage syslog-msg)))
      (is (= Facility/KERN
             (.getFacility ^SyslogMessage syslog-msg))))))

(defn fake-tcp-syslog
  [results _transport]
  (proxy [TcpSyslogMessageSender] []
    (sendMessage [msg] (s/put! results msg))))

(deftest ->syslog!-tests
  (testing "complete example"
    (let [results (s/stream)
          inputs (s/stream)
          conn-opts {:host "localhost"
                     ;; Do we parse port -> int correctly?
                     :port "1895"
                     ;; tls transport has the most interesting behavior
                     :transport :tls
                     :message-format :rfc-5424}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}
          ;; Above, KERN overrides unclogged default, which is USER.
          ;; This is meant to test that message details override syslog
          ;; client instance defaults override our package defaults.
          message-details {:message "only in message"
                           :message-id "only in message"}]
      (with-redefs [unclogged.core/make-syslog (partial fake-tcp-syslog results)]
        (c/->syslog! inputs conn-opts defaults)
        (s/put! inputs message-details)
        (let [syslog-message @(s/take! results)]
          (is (= "only in message"
                 (.toString ^CharArrayWriter (.getMsg syslog-message))))
          (is (= "only in message"
                 (.getMsgId syslog-message)))
          (is (= "ditka"
                 (.getAppName syslog-message)))
          (is (= "dabears"
                 (.getHostname syslog-message)))
          (is (= "89"
                 (.getProcId syslog-message)))
          (is (= Facility/KERN
                 (.getFacility syslog-message))) ;; instance-default
          (is (= Severity/INFORMATIONAL
                 (.getSeverity syslog-message)))) ;; unclogged default
        (let [{:keys [syslog stream]} (c/->syslog! inputs conn-opts defaults)]
          (is (some? syslog))
          (is (= "localhost" (.getSyslogServerHostname syslog)))
          (is (= 1895 (.getSyslogServerPort syslog)))
          (is (= MessageFormat/RFC_5424 (.getMessageFormat syslog)))
          (is (.isSsl syslog))))))
  (testing "defaults, no integer port parsing"
    (let [inputs (s/stream)
          conn-opts {:host "localhost"
                     :port 1895}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}]
      (c/->syslog! inputs conn-opts defaults)
      (let [{:keys [syslog stream]} (c/->syslog! inputs conn-opts defaults)]
        (is (some? syslog))
        (is (instance? TcpSyslogMessageSender syslog))
        (is (= "localhost" (.getSyslogServerHostname syslog)))
        (is (= 1895 (.getSyslogServerPort syslog)))
        (is (= MessageFormat/RFC_3164 (.getMessageFormat syslog))
            "default message format is RFC3164")
        (is (.isSsl syslog)
            "default transport is TLS"))))
  (testing "TCP, explicit 5424"
    (let [inputs (s/stream)
          conn-opts {:host "localhost"
                     :port 1895
                     :transport :tcp
                     :message-format :rfc-5424}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}]
      (c/->syslog! inputs conn-opts defaults)
      (let [{:keys [syslog stream]} (c/->syslog! inputs conn-opts defaults)]
        (is (some? syslog))
        (is (instance? TcpSyslogMessageSender syslog))
        (is (= "localhost" (.getSyslogServerHostname syslog)))
        (is (= 1895 (.getSyslogServerPort syslog)))
        (is (= MessageFormat/RFC_5424 (.getMessageFormat syslog)))
        (is (not (.isSsl syslog))))))
  (testing "TLS, explicit 5424"
    (let [inputs (s/stream)
          conn-opts {:host "localhost"
                     :port 1895
                     :transport :tls
                     :message-format :rfc-5424}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}]
      (c/->syslog! inputs conn-opts defaults)
      (let [{:keys [syslog stream]} (c/->syslog! inputs conn-opts defaults)]
        (is (some? syslog))
        (is (instance? TcpSyslogMessageSender syslog))
        (is (= "localhost" (.getSyslogServerHostname syslog)))
        (is (= 1895 (.getSyslogServerPort syslog)))
        (is (= MessageFormat/RFC_5424 (.getMessageFormat syslog)))
        (is (.isSsl syslog)))))
  (testing "SSL is just an alias for TLS, explicit 5424"
    (let [inputs (s/stream)
          conn-opts {:host "localhost"
                     :port 1895
                     :transport :ssl
                     :message-format :rfc-5424}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}]
      (c/->syslog! inputs conn-opts defaults)
      (let [{:keys [syslog stream]} (c/->syslog! inputs conn-opts defaults)]
        (is (some? syslog))
        (is (instance? TcpSyslogMessageSender syslog))
        (is (= "localhost" (.getSyslogServerHostname syslog)))
        (is (= 1895 (.getSyslogServerPort syslog)))
        (is (= MessageFormat/RFC_5424 (.getMessageFormat syslog)))
        (is (.isSsl syslog)))))
  (testing "Explicit UDP"
    (let [inputs (s/stream)
          conn-opts {:host "localhost"
                     :port 1895
                     :transport :udp}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}]
      (let [{:keys [syslog stream]} (c/->syslog! inputs conn-opts defaults)]
        (is (some? syslog))
        (is (instance? UdpSyslogMessageSender syslog))
        (is (= "localhost" (.getSyslogServerHostname syslog)))
        (is (= 1895 (.getSyslogServerPort syslog)))
        (is (thrown? IllegalArgumentException (.isSsl syslog)))))))

(deftest syslog-sink-tests
  (testing "Basic sink setup"
    (let [results (s/stream)
          conn-opts {;; getting hostname later will call InetAddress's
                     ;; getByName, which tries to resolve. So, the host has to
                     ;; be resolvable, or the tests fail. I wanted "halas".
                     :host "localhost"
                     :port 1895
                     ;; we use the tls transport because that has
                     ;; the most interesting behavior
                     :transport :tls
                     :message-format :rfc-5424}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}
          ;; Above, KERN overrides unclogged default, which is USER.
          ;; This is meant to test that message details override syslog
          ;; client instance defaults override our package defaults.
          message-details {:message "only in message"
                           :message-id "only in message"}]
      (with-redefs [unclogged.core/make-syslog (partial fake-tcp-syslog results)]
        (let [{:keys [syslog stream]} (c/syslog-sink conn-opts defaults)]
          (s/put! stream message-details)
          (let [syslog-message @(s/take! results)]
            (is (= "only in message"
                   (.toString ^CharArrayWriter (.getMsg syslog-message))))
            (is (= "only in message"
                   (.getMsgId syslog-message)))
            (is (= "ditka"
                   (.getAppName syslog-message)))
            (is (= "dabears"
                   (.getHostname syslog-message)))
            (is (= "89"
                   (.getProcId syslog-message)))
            (is (= Facility/KERN
                   (.getFacility syslog-message))) ;; instance-default
            (is (= Severity/INFORMATIONAL
                   (.getSeverity syslog-message))) ;; unclogged default

            (is (some? syslog))
            (is (= "localhost" (.getSyslogServerHostname syslog)))
            (is (= 1895 (.getSyslogServerPort syslog)))
            (is (= MessageFormat/RFC_5424 (.getMessageFormat syslog)))
            (is (.isSsl syslog)))))))
  (testing "Sink setup with customizable buffer size"
    (let [buffer-size 3
          results (s/stream buffer-size)
          conn-opts {:host "localhost"
                     :port 1895
                     :buffer-size buffer-size
                     :transport :tls
                     :message-format :rfc-5424}
          defaults {:hostname "dabears"
                    :app-name "ditka"
                    :process-id 89
                    :facility Facility/KERN}]
      (with-redefs [unclogged.core/make-syslog (partial fake-tcp-syslog results)]
        (let [{:keys [stream]} (c/syslog-sink conn-opts defaults)]
          (is (= buffer-size (:buffer-capacity (s/description stream))))

          (dotimes [n buffer-size] (s/put! stream (compose-message n)))
          (is (= buffer-size (:buffer-size (s/description results))))
          (is (= 0 (:pending-puts (s/description results))))

          (s/put! stream (compose-message 4))
          (is (= buffer-size (:buffer-size (s/description results))))
          (is (= 1 (:pending-puts (s/description results)))))))))


src/unclogged/core.clj
(ns unclogged.core
  (:require
   [manifold.stream :as s]
   [taoensso.timbre :refer [spy]])
  (:import
   [com.cloudbees.syslog.sender TcpSyslogMessageSender UdpSyslogMessageSender]
   [com.cloudbees.syslog Facility Severity MessageFormat SyslogMessage]))

(defn ^:private parse-facility
  [x]
  (cond
    (instance? Facility x) x
    (keyword? x) (parse-facility (name x))
    (string? x) (Facility/fromLabel (.toUpperCase ^String x))
    (number? x) (Facility/fromNumericalCode x)))

(def ^:private severity-aliases
  {"INFO" Severity/INFORMATIONAL
   "WARN" Severity/WARNING
   "ERR" Severity/ERROR})

(defn ^:private parse-severity
  [x]
  (cond
    (instance? Severity x) x
    (number? x) (Severity/fromNumericalCode x)
    :else (let [x (.toUpperCase ^String (name x))]
            (or (get severity-aliases x)
                (Severity/fromLabel x)))))

(defn ^:private parse-message-format
  [s]
  (case (re-find #"\d+" (name s))
    "3164" MessageFormat/RFC_3164
    "5424" MessageFormat/RFC_5424))

(defn ^:private ->syslog-msg
  "Turns defaults + message map into a SyslogMessage."
  [defaults message]
  (let [{:keys [message
                message-id
                app-name
                hostname
                process-id
                facility
                severity]}
        (merge defaults message)]
    (doto (SyslogMessage.)
      (.withMsg (str message))
      (.withMsgId message-id)
      (.withAppName app-name)
      (.withHostname hostname)
      (.withProcId (str process-id))
      (.withFacility (parse-facility facility))
      (.withSeverity (parse-severity severity)))))

(defn ^:private make-syslog
  "Creates an instance.

  This fn exists because it's way easier to redef a Clojure fn than it
  is to mock a Java class."
  [transport]
  (case transport
    :udp (UdpSyslogMessageSender.)
    :tcp (TcpSyslogMessageSender.)
    ;; TLS is enabled in configured-syslog.
    :tls (TcpSyslogMessageSender.)
    :ssl (TcpSyslogMessageSender.)))

(defn ^:private configured-syslog
  "Creates a configured syslog instance.

  Please note that this currently does not set any message defaults;
  all values are always set on all syslog messages; this may be a
  small performance hit, but helps inspectability. Ideally, our syslog
  implementation wouldn't check for that at all, and decomplect the
  concerns of providing defaults with the actual send-syslog-messages
  bits."
  [conn-opts]
  (let [{:keys [host port message-format transport]
         :or {transport :tls}} conn-opts
        syslog (make-syslog transport)
        port (if (string? port)
               (Integer/parseInt port)
               port)]
    (.setSyslogServerHostname syslog host)
    (when port
      (.setSyslogServerPort syslog port))
    (when message-format
      (.setMessageFormat syslog (parse-message-format message-format)))
    (when (#{:tls :ssl} transport)
      (.setSsl syslog true))
    syslog))

(def system-defaults
  "Unclogged's default chocies for syslog messages."
  {:severity Severity/INFORMATIONAL
   :facility Facility/USER})

(defn ->syslog!
  "Consumes elems on source and sends them to syslog as specified by
  the connection map & defaults.

  Values in message maps override values in the defaults; values in
  the defaults override unclogged's default-defaults. This prevents
  NullPointerExceptions when you forget to provide a facility or
  severity, for example.

  Returns a map containing the syslog client object under the :syslog
  key. This is only provided for inspection; mutating that object is
  not guaranteed to have desired effects. It also contains the input
  stream under the :stream key."
  [source conn-opts defaults]
  (let [->syslog-msg (partial ->syslog-msg (merge system-defaults defaults))
        syslog (configured-syslog conn-opts)
        send! (fn [msg] (.sendMessage syslog (->syslog-msg (spy msg))))]
    (s/consume send! source)
    {:syslog syslog
     :stream source}))

(defn syslog-sink
  "Builds a Manifold sink (stream) where you can dump information
  you'd like to send to syslog. Takes a connection map and some
  message defaults.

  Returns a map that contains at least the built stream under
  the :stream key.

  If you already have a manifold stream, see ->syslog!."
  [conn-opts defaults]
  (let [sink (s/stream (conn-opts :buffer-size 0))]
    (->syslog! sink conn-opts defaults)))
